src/video.cpp

```diff
@@ -349,36 +350,116 @@ bool VideoSystem::HandleEvent()
+				case SDLK_RIGHT:
+					return HandleKeyInput(WMKC_CURSOR_RIGHT, this->modifier_state);
+
+				case SDLK_KP_2:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_DOWN:
+					return HandleKeyInput(WMKC_CURSOR_DOWN, this->modifier_state);
+
+				case SDLK_KP_9:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_PAGEUP:
+					return HandleKeyInput(WMKC_CURSOR_PAGEUP, this->modifier_state);
+
+				case SDLK_KP_3:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_PAGEDOWN:
+					return HandleKeyInput(WMKC_CURSOR_PAGEDOWN, this->modifier_state);
+
+				case SDLK_KP_7:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_HOME:
+					return HandleKeyInput(WMKC_CURSOR_HOME, this->modifier_state);
+
+				case SDLK_KP_1:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_END:
+					return HandleKeyInput(WMKC_CURSOR_END, this->modifier_state);
+
+				case SDLK_KP_PERIOD:
+					if ((event.key.keysym.mod & KMOD_NUM) != 0) break;
+				case SDLK_DELETE:
+					return HandleKeyInput(WMKC_DELETE, this->modifier_state);
 
 				case SDLK_RETURN:
 				case SDLK_RETURN2:
-				case SDLK_KP_ENTER:  return HandleKeyInput(WMKC_CONFIRM);
+				case SDLK_KP_ENTER:
+					return HandleKeyInput(WMKC_CONFIRM, this->modifier_state);
+
+				case SDLK_ESCAPE:
+					return HandleKeyInput(WMKC_CANCEL, this->modifier_state);
+
+				case SDLK_BACKSPACE:
+					return HandleKeyInput(WMKC_BACKSPACE, this->modifier_state);
+
+				case SDLK_LCTRL:
+				case SDLK_RCTRL:
+					this->modifier_state |= WMKM_CTRL;
+					return true;
+
+				case SDLK_LALT:
+				case SDLK_RALT:
+				case SDLK_LGUI:
+				case SDLK_RGUI:
+					this->modifier_state |= WMKM_ALT;
+					return true;
 
 				case SDLK_LSHIFT:
 				case SDLK_RSHIFT:
+					this->modifier_state |= WMKM_SHIFT;
+
 					_game_control.action_test_mode = true;
 					return true;
 
-				default:             return false;
+				default:
+					/* Text input events with modifiers may or may not be recognized as SDL_TEXTINPUT, so we need to convert them manually.
+					 * All keysyms that correspond to an ASCII character have the same integer value as this character;
+					 * all others are larger than the largest valid ASCII character (which is 0x7F).
+					 */
+					if (this->modifier_state != WMKM_NONE && event.key.keysym.sym <= 0x7F) {
+						std::string text;
+						text.push_back(event.key.keysym.sym);
+						return HandleKeyInput(WMKC_SYMBOL, this->modifier_state, text);
+					}
+
+					return false;
 			}
 
 		case SDL_KEYUP:
 			switch (event.key.keysym.sym) {
+				case SDLK_LCTRL:
+				case SDLK_RCTRL:
+					this->modifier_state &= ~WMKM_CTRL;
+					return true;
+				case SDLK_LALT:
+				case SDLK_RALT:
+				case SDLK_LGUI:
+				case SDLK_RGUI:
+					this->modifier_state &= ~WMKM_ALT;
+					return true;
 				case SDLK_LSHIFT:
 				case SDLK_RSHIFT:
+					this->modifier_state &= ~WMKM_SHIFT;
+
```

Comment:
Why an empty line here?

I'd expect empty lines below a code so it becomes easier to see which cases belong together.